src/rust-util/pelikan/src/protocol/admin.rs

```diff
@@ -13,11 +13,146 @@
 // limitations under the License.
 
 use super::*;
-use pelikan_sys::protocol::admin::{request, response};
+use ccommon::buf::OwnedBuf;
+use ccommon_sys::buf;
+use pelikan_sys::protocol::admin::{
+    admin_compose_req, admin_compose_rsp, admin_parse_req, admin_request_reset,
+    admin_response_reset, request, response, COMPOSE_ENOMEM, COMPOSE_EOVERSIZED, PARSE_EINVALID,
+    PARSE_EUNFIN, PARSE_OK, REQ_QUIT,
+};
+
+use std::convert::Infallible;
+use std::error::Error;
+use std::fmt::{self, Display, Formatter};
 
 pub enum AdminProtocol {}
 
+#[derive(Default)]
+#[repr(transparent)]
+pub struct Request(pub request);
+
+#[derive(Default)]
+#[repr(transparent)]
+pub struct Response(pub response);
+
+#[derive(Debug)]
+pub enum ParseError {
+    Unfinished,
+    Invalid,
+    Other,
+}
+
+#[derive(Debug)]
+pub enum ComposeError {
+    NoMem,
+    Oversized,
+    Other,
+}
+
 impl Protocol for AdminProtocol {
-    type Request = request;
-    type Response = response;
+    type Request = Request;
+    type Response = Response;
+}
+
+impl Serializable for Request {
+    type ParseError = ParseError;
+    type ComposeError = ComposeError;
+
+    fn reset(&mut self) {
+        unsafe { admin_request_reset(&mut self.0 as *mut _) }
+    }
+
+    fn parse(&mut self, buf: &mut OwnedBuf) -> Result<(), Self::ParseError> {
+        unsafe {
+            let status = admin_parse_req(&mut self.0 as *mut _, buf.as_mut_ptr());
+
+            match status {
+                PARSE_OK => (),
+                PARSE_EUNFIN => return Err(ParseError::Unfinished),
+                PARSE_EINVALID => return Err(ParseError::Invalid),
+                _ => return Err(ParseError::Other),
+            }
+
+            Ok(())
+        }
+    }
+    fn compose(&self, buf: &mut OwnedBuf) -> Result<usize, Self::ComposeError> {
+        unsafe {
+            let status = admin_compose_req(
+                // Not sure what's the proper pattern here
+                buf as *mut OwnedBuf as *mut *mut buf,
+                &self.0 as *const request,
+            );
+
+            match status {
+                amt if amt >= 0 => Ok(amt as usize),
+                COMPOSE_ENOMEM => Err(ComposeError::NoMem),
+                COMPOSE_EOVERSIZED => Err(ComposeError::Oversized),
+                _ => Err(ComposeError::Other),
+            }
+        }
+    }
+}
+impl Serializable for Response {
+    type ParseError = Infallible;
+    type ComposeError = ComposeError;
+
+    fn reset(&mut self) {
+        unsafe { admin_response_reset(&mut self.0 as *mut _) }
+    }
+
+    fn parse(&mut self, _: &mut OwnedBuf) -> Result<(), Self::ParseError> {
+        unimplemented!()
+    }
+    fn compose(&self, buf: &mut OwnedBuf) -> Result<usize, Self::ComposeError> {
```

Comment:
Fixed - note that rustfmt doesn't add spaces between functions in the default config.

Maybe we should add a `rustfmt.toml` to the repo so that it's done by default. I'm not sure if rustfmt supports it but it would be useful if it does.

Reply:
Let's look into that separately